docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,71 @@
+
+Methods may be either 'sync' or 'async'. Async methods have a special signature encoding, described in [### I.8.6.1.5 Method signatures].
+
+Sync methods are all other methods.
+
+Unlike sync methods, async methods support suspension. Suspension allows async methods to yield control flow back to their caller at certain well-defined suspension points, and resume execution of the remaining method at a later time or location, potentially on another thread.
+
+Async methods support the following suspension points:
+
+* Calling another async method. No special instructions need to be provided. If the callee suspends, the caller will suspend as well.
+* Using a library in the .NET corelib to "await" an "INotifyCompletion" type. The signatures of these methods shall be:
+  ```C#
+  // public static async2 Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  // public static async2 Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  ```
+
+Each of the above methods will have semantics analogous to the current AsyncTaskMethodBuilder.AwaitOnCompleted/AwaitUnsafeOnCompleted methods. After calling this method, in can be presumed that the task has completed.
+
+Async methods have the following restrictions:
+* Usage of the `localloc` and instruction is forbidden
+* The `lodloca` and `ldarga` instructions are redefined to return managed pointers instead of pointers.
+* Pinning locals may not be created
+* The `tail` prefix is forbidden
+
+Suspension points may not appear in exception handling blocks.
+
+All async methods effectively have two entry points, or signatures. The first signature is the one present in the above code: a modopt before the return type. The second signature is a "Task-equivalent signature", described in further detail in [I.8.6.1.5 Method signatures].
+
+Async methods have a special calling convention and may not be called directly outside of other async methods. To call an async method from a sync method, callers must use the second "Task-equivalent signature".
+
+Callers may retrieve a Task-equivalent return type from an async method via calling the "Task-equivalent signature". This functionality is available in both sync and async methods.
+
+### I.8.6.1.5 Method signatures
+
+The list of relevant components is augmented to include sync vs. async method types. Async methods have some additions to normal signature compatibility.
+
+Async MethodDef entries implicitly create two member definitions: one explicit, primary definiton, and a second implicit, runtime-generated definition.
```

Comment:
Can you comment on the impact to binary compatibility here? When the compiler generates a `.call`, what signature do we expect it to provide? Either?

Reply:
I think the idea is that if you're in an `async2` context, you emit the `async2` signature to call (so that way you get optimised behaviour when the method is implemented with `async2`), otherwise the task based one. The runtime generates the one that was not provided.

Reply:
The precise idea is that you emit a call to _either_ signature, depending on what you need. So if you're in a (runtime) async method and want to await another async method, you emit a call to the "primary" signature. If you want to convert that method to a task, either because in source code there's no await, or because you're not in an async method, you call the "Task-equivalent" signature.

Reply:
When an async2 method calls another async2 method directly it is roughly the same as calling a Task-equivalent signature and immediately `await`ing the result, just optimized to bypass the Task API. 
In theory such call can be inlined or tail-call optimized, or just performed much faster since Task does not stand in the way.

One reason to call Task-equivalent signature is if the caller actually wants to use Task APIs like `WaitAll`, `ContinueWith`.  
This is a plausible scenario for both ordinary and runtime-async callers.

For the ordinary methods the Task-based signature is the only one that they can call. A runtime async method can call either of the signatures. 
Direct call is the "magic" call that propagates the asynchrony, while call via Task API is the old boring call which will get you a Task.

> Can you comment on the impact to binary compatibility here? When the compiler generates a .call, what signature do we expect it to provide? Either?  

Presumably the preexisting `.call` was calling async-v1 method and expecting a Task result, thus it should continue calling the Task-based signature. 
Effectively the caller does not need to change at all if the calee is replaced with a runtime async equivalent.

Reply:
Interesting, but this then raises another question for me: does this mean that it becomes a _binary breaking_ change to take an `async2` method and make it not actually `async`? IE, if I change it to just a `Task`-returning method, the compiler is no longer going to emit the "primary" async runtime signature; does that mean that an `async2` method in another binary that was calling it in the "primary" form breaks?

Reply:
Talked with @agocke this morning: while the rules as written today may not adequately convey this, the intent is that the above scenario is _not_ a breaking change. I think it likely that the C# compiler will want to take the emit strategy of: if the runtime feature flag exists, always emit directly-awaited task-or-valuetask-returning methods in the new primary async signature form, _regardless_ of whether the compiler sees the new primary async signature in the reference or not. That way, if the reference updates (or if this is a complex multi-targeting scenario where the final implementation has the new signature, but the reference the compiler saw beforehand doesn't), things will seamlessly get faster, without having to go through multiple recompiles for an async signature change to propagate through the system.

Reply:
Addendum has been added to basically say that any sync method that would correspond to an async method also has an async definition. It's not particularly elegant, but ideally that will be cleaned up.